opm/grid/cpgrid/CpGrid.cpp

```diff
@@ -2485,143 +2485,235 @@ bool CpGrid::adapt(const std::vector<std::array<int,3>>& cells_per_dim_vec,
+    const auto& globalIdSet = this->globalIdSet();
+
+    parentToFirstChildGlobalIds.resize(data[0]->size(0), -1); // Initialize with -1 (invalid value, for non parent cells).
+
+    const auto& elements = Dune::elements(levelGridView(0));
+    for (const auto& element : elements) {
+        const auto& [level, children] = parentToChildrenBeforeLoadBalance[element.index()];
 
-    return preAdapt();
+        if (!children.empty()) {
+            const auto& levelData = *data[level];
+            const auto& first_child = Dune::cpgrid::Entity<0>(levelData, children[0], true);
+
+            // Rewrite parent global id entry with first child global id
+            parentToFirstChildGlobalIds[globalIdSet.id(element)] = globalIdSet.id(first_child);
+        }
+    }
+#endif
 }
 
-void CpGrid::postAdapt()
+void CpGrid::syncDistributedGlobalCellIds()
 {
-    // - Resize with the new amount of cells on the leaf grid view
-    // - Set marks equal to zero (representing 'doing nothing')
-    current_view_data_ -> postAdapt();
+#if HAVE_MPI
+    std::vector<int> parentToFirstChildGlobalIds;
+    getFirstChildGlobalIds(parentToFirstChildGlobalIds);
+
+    auto size = comm().max(parentToFirstChildGlobalIds.size());
+
+    switchToDistributedView();
+
+    parentToFirstChildGlobalIds.resize(size);
+    comm().broadcast(parentToFirstChildGlobalIds.data(), parentToFirstChildGlobalIds.size(), 0);
+
+    const int maxLevel = this->maxLevel();
+
+    // Preallocate syncCellIds
+    std::vector<std::vector<int>> syncCellIds(maxLevel);
+    for (int level = 1; level <= maxLevel; ++level) {
+        syncCellIds[level-1].resize(currentData()[level]->size(0));
+    }
+
+    const auto& globalIdSet = this->globalIdSet();
+
+    // Populate for interior cells
+    for (int level = 1; level <= maxLevel; ++level) {
+        const auto& elements = Dune::elements(levelGridView(level));
+        for (const auto& element : elements) {
+            const int parent_globalId = globalIdSet.id(element.father());
+            const int idx_in_parent = element.getIdxInParentCell();
+            const int first_child_id = parentToFirstChildGlobalIds[parent_globalId];
+            const int new_elem_globalId = first_child_id + idx_in_parent;
+
+            syncCellIds[element.level()-1][element.index()] = new_elem_globalId;
+        }
+    }
+
+    // Re-assign new cell global ids for all refined level grids
+    std::vector<int> faceIds; // empty for all
+    for (int level = 1; level <= maxLevel; ++level) {
+        if(currentData()[level]->size(0)) { // Check if LGR is active in currect process.
+            auto vertexIds = currentData()[level]->global_id_set_-> getMapping<3>();
+            currentData()[level]->global_id_set_->swap(syncCellIds[level-1],
+                                                       faceIds,
+                                                       vertexIds);
+
+            populateCellIndexSetRefinedGrid(level);
```

Comment:
Sorry, I missed this one during the first review. It is a bit unclear to me how this behaves:

- The method does not clear the interface before adding. I am not sure what happens with our approach, might be one of these
  - We have multiple entries with the same global id
  - We only have one, but cannot be sure which one of the two wins.
- RemoteIndices using the ParallelIndexSet will be broken. Hence we will need to rebuild/update them. Not sure whether there some and whether we use those later.

I guess the easy fix is to clear the index set in the function before adding and then maybe rebuild remote indices.
In the long run this should have its own way of doing things as we actually need to do less then in populateCellIndexSetRefinedGrid.

Reply:
Thanks for noticing, indeed, the size of cell index set for levels was not correct (since there was no clearing of it before the assignment of new sync ids). This [commit](https://github.com/OPM/opm-grid/pull/854/commits/0f88e5666b0cd014176db09c616f4074e91ec1a2) fixes it.